File: built-in plugins dsp/mb-loudness-max.cpp
/**
 * MB Loudness Maximizer
 * Category : effect
 * Type     : limiter
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Intelligent loudness maximizer with LUFS targeting
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_LOUDNESS_MAX_H
#define MB_LOUDNESS_MAX_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbLoudnessMax : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-loudness-max";
    static constexpr const char* PLUGIN_NAME    = "MB Loudness Maximizer";
    static constexpr const char* PLUGIN_TYPE    = "limiter";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float target = -14f;  // range [-24, -6]
    float ceiling = -0.3f;  // range [-3, 0]
    float speed = 0.5f;  // range [0, 1]
    float character = 0.5f;  // range [0, 1]
    };

    MbLoudnessMax() = default;
    ~MbLoudnessMax() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.target = std::clamp(params.target, -24f, -6f);
        params.ceiling = std::clamp(params.ceiling, -3f, 0f);
        params.speed = std::clamp(params.speed, 0f, 1f);
        params.character = std::clamp(params.character, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Loudness Maximizer
        return input;
    }
};

#endif // MB_LOUDNESS_MAX_H

File: built-in plugins dsp/mb-loudness-max.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbLoudnessMaxPlugin: PluginDefinition = { id: 'mb-loudness-max', slug: 'mb-loudness-max', name: 'MB Loudness Maximizer', category: 'effect', type: 'limiter' as any, version: '1.0.0', description: 'Intelligent loudness maximizer with LUFS targeting', author: 'Max Booster', grade: 'A', parameters: [{ id: 'target', name: 'Target LUFS', type: 'float', defaultValue: -14, minValue: -24, maxValue: -6, automatable: true }, { id: 'ceiling', name: 'Ceiling', type: 'float', defaultValue: -0.3, minValue: -3, maxValue: 0, automatable: true }, { id: 'speed', name: 'Speed', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'character', name: 'Character', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { target: -14, ceiling: -0.3, speed: 0.5, character: 0.5 } };

export default MbLoudnessMaxPlugin;
